tests/src/test/java/alluxio/BaseIntegrationTest.java

```diff
@@ -21,41 +23,65 @@
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
+import java.util.Collections;
+import java.util.List;
 
 /**
  * Base class used for specifying the maximum time a test should run for.
  */
 public abstract class BaseIntegrationTest {
-
   @Rule
-  public Timeout mGlobalTimeout = Timeout.millis(Constants.MAX_TEST_DURATION_MS);
+  public RuleChain mRules = createRuleChain();
 
-  @Rule
-  public TestWatcher mWatcher = new TestWatcher() {
-    // When tests fail, save the logs.
-    protected void failed(Throwable t, Description description) {
-      try {
-        Files.copy(Paths.get(Constants.TESTS_LOG),
-            Paths.get(Constants.TEST_LOG_DIR, String.format("%s-%s.log",
-                description.getClassName(), description.getMethodName())),
-            StandardCopyOption.REPLACE_EXISTING);
-      } catch (IOException e) {
-        if (t != null) {
-          t.addSuppressed(e);
-        } else {
-          throw new RuntimeException(e);
+  private RuleChain createRuleChain() {
+    RuleChain chain = RuleChain.emptyRuleChain();
+    chain = chain.around(logHandler());
+    for (TestRule rule : rules()) {
+      chain = chain.around(rule);
+    }
+    // Make Timeout the innermost rule so that other rules run in the main thread. Everything inside
+    // Timeout is executed in a new thread.
+    chain = chain.around(Timeout.millis(Constants.MAX_TEST_DURATION_MS));
+    return chain;
+  }
+
+  private TestWatcher logHandler() {
+    return new TestWatcher() {
+      // When tests fail, save the logs.
+      protected void failed(Throwable t, Description description) {
+        try {
+          Files.copy(Paths.get(Constants.TESTS_LOG),
+              Paths.get(Constants.TEST_LOG_DIR, String.format("%s-%s.log",
+                  description.getClassName(), description.getMethodName())),
+              StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
+          if (t != null) {
+            t.addSuppressed(e);
```

Comment:
Do we need to throw `t` here?

Reply:
no, the base class handles exceptions and just notifies the subclass about them - that's why there's no `throws Exception` in the signature for `failed`